File: code/ch14/14.3.3_240.search-a-2d-matrix-ii.java
class Solution {
  private boolean binarySearch(int[][] matrix, int target, int start, boolean vertical) {
      int lo = start;
      int hi = vertical ? matrix[0].length-1 : matrix.length-1;

      while (hi >= lo) {
          int mid = (lo + hi)/2;
          if (vertical) { // searching a column
              if (matrix[start][mid] < target) {
                  lo = mid + 1;
              } else if (matrix[start][mid] > target) {
                  hi = mid - 1;
              } else {
                  return true;
              }
          } else { // searching a row
              if (matrix[mid][start] < target) {
                  lo = mid + 1;
              } else if (matrix[mid][start] > target) {
                  hi = mid - 1;
              } else {
                  return true;
              }
          }
      }

      return false;
  }

  public boolean searchMatrix(int[][] matrix, int target) {
      // an empty matrix obviously does not contain `target`
      if (matrix == null || matrix.length == 0) {
          return false;
      }

      // iterate over matrix diagonals
      int shorterDim = Math.min(matrix.length, matrix[0].length);
      for (int i = 0; i < shorterDim; i++) {
          boolean verticalFound = binarySearch(matrix, target, i, true);
          boolean horizontalFound = binarySearch(matrix, target, i, false);
          if (verticalFound || horizontalFound) {
              return true;
          }
      }
      
      return false; 
  }
}

File: code/ch14/14.3.3_240.search-a-2d-matrix-ii.py
from typing import List
class Solution:
    def binary_search(self, matrix: List[List[int]], target: int, start: int, vertical: bool) -> bool:
        lo = start
        hi = len(matrix[0]) - 1 if vertical else len(matrix) - 1

        while lo <= hi:
            mid = lo + (hi - lo) // 2
            if vertical:
                if matrix[start][mid] < target:
                    lo = mid + 1
                elif matrix[start][mid] > target:
                    hi = mid - 1
                else:
                    return True
            else:
                if matrix[mid][start] < target:
                    lo = mid + 1
                elif matrix[mid][start] > target:
                    hi = mid - 1
                else:
                    return True

        return False

    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix:
            return False

        minLen = min(len(matrix), len(matrix[0]))
        for i in range(minLen):
            vertical_found = self.binary_search(matrix, target, i, True)
            horizontal_found = self.binary_search(matrix, target, i, False)
            if vertical_found or horizontal_found:
                return True

        return False
